### PR TITLE
GPT function calling

### PR DESCRIPTION
**Pull request by ohadschn**
Should greatly increase the chance of pure JSON response, which in turn should greatly reduce our reliance on the current (easily broken) mechanism JSON extraction from free-text: https://platform.openai.com/docs/guides/gpt/function-calling

* Sample run: https://github.com/ohadschn/ConnectPortal/actions/runs/6684655429/job/18162157181
* Sample comment: https://github.com/ohadschn/ConnectPortal/issues/14#issuecomment-1784167671

Also, constructs the schema provided to GPT by parsing the new initiative issue form, so any new parameter we add there will automatically be known to GPT.

PS here's how I found out about it (now posted an answer following one of the comments): https://stackoverflow.com/questions/76185628/how-to-prompt-chatgpt-api-to-give-completely-machine-readable-responses-without

### CODE REVIEW

**Comment by 4tal** [comment]
url and website keys are different. website is a website and url is the main url of the project. It can be whatsapp, facebook, website, and so on.

**Comment by ohadschn** [comment]
> website

> url and website keys are different. website is a website and url is the main url of the project. It can be whatsapp, facebook, website, and so on.

I see, yet another reason for https://github.com/4tals/LinksForIsrael/issues/293

**Review by 4tal — APPROVED** [review]

File: .github/ISSUE_TEMPLATE/new_initiative.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ body:
       required: true
 
   - type: input
-    id: short-description
+    id: shortDescription
```
[the remark this answers is not in the document]
Comment (ohadschn) [reply .github/ISSUE_TEMPLATE/new_initiative.yml:15]: These should now match the keys in our JSON schema, as they are imported by the Python script (prevents duplication)

File: scripts/gptGenerateJsonFromIssue.py [hunk 2]
```diff
@@ -2,120 +2,116 @@
...
             {"role": "user", "content": query}
         ]
-        response = openai.ChatCompletion.create(
-            model=self.model,
-            messages=conversation
-        )
-        answer = response['choices'][0]['message']['content']
-        return answer
-
-    def query_via_rest(self, text_string: str) -> str:
```
[the remark this answers is not in the document]
Comment (ohadschn) [reply scripts/gptGenerateJsonFromIssue.py:74]: I don't see why - under the covers the SDK also issues REST requests, except it potentially wraps them with retries, logging, possibly multiple requests for a single logical operation, etc.

TLDR the whole point of a good SDK is that it abstracts the underlying REST API for you (otherwise we wouldn't ever need SDKs)

File: scripts/gptGenerateJsonFromIssue.py [hunk 3]
```diff
@@ -2,120 +2,116 @@
...
-            response = assistant.query_via_api(initiative_details)
-            output_response(response)
-    except Exception as e:
-        logging.error(f"An error occurred: {e}")
+    gpt_response = GPTAssistant().ask_gpt_for_initiative_json(os.environ["ISSUE_BODY"])
+    
+    respose_output_file = os.environ["GITHUB_WORKSPACE"] + "/gpt-auto-comment.output"
+    logging.info(f"Writing GPT Response to {respose_output_file}:\n{gpt_response}")
+
+    with open(respose_output_file, "w", encoding="utf-8") as output:
```
[the remark this answers is not in the document]
Comment (ohadschn) [reply scripts/gptGenerateJsonFromIssue.py:116]: Sorry, what race condition?
This is all synchronous...
